src/deploy/functions/runtimes/golang/index.ts

```diff
@@ -56,113 +63,119 @@ export async function tryCreateDelegate(
-      continue;
-    }
-
-    if (line.trim()) {
-      logger.debug("Don't know how to handle line", line, "in mod.go");
-    }
-  }
-
-  if (!module.module) {
-    throw new FirebaseError("Module has no name");
-  }
-  if (!module.version) {
-    throw new FirebaseError(`Module ${module.module} has no go version`);
-  }
-
-  return module;
-}
-
 export class Delegate {
   public readonly name = "golang";
 
   constructor(
     private readonly projectId: string,
     private readonly sourceDir: string,
     public readonly runtime: runtimes.Runtime,
-    private readonly module: Module
+    private readonly module: gomod.Module
   ) {}
   validate(): Promise<void> {
-    throw new FirebaseError("Cannot yet analyze Go source code");
+    return Promise.resolve();
   }
 
-  build(): Promise<void> {
-    const res = spawn.sync("go", ["build"], {
+  async build(): Promise<void> {
+    try {
+      await promisify(fs.mkdir)(path.join(this.sourceDir, "autogen"));
+    } catch (err) {
+      if (!/EEXIST/.exec(err?.message)) {
+        throw new FirebaseError("Failed to create codegen directory", { children: [err] });
+      }
+    }
+    const genBinary = spawn.sync("go", ["run", FUNCTIONS_CODEGEN, this.module.module], {
       cwd: this.sourceDir,
-      stdio: "inherit",
+      stdio: [/* stdin=*/ "ignore", /* stdout=*/ "pipe", /* stderr=*/ "pipe"],
     });
-    if (res.error) {
-      logger.debug("Got error running go build", res);
-      throw new FirebaseError("Failed to build functions source", { children: [res.error] });
+    if (genBinary.status != 0) {
+      throw new FirebaseError("Failed to run codegen", {
+        children: [new Error(genBinary.stderr.toString())],
+      });
+    }
+    await promisify(fs.writeFile)(
+      path.join(this.sourceDir, "autogen", "main.go"),
+      genBinary.stdout
+    );
+
+    // Now that we've created main.go, we will have new dependencies on the packages pulled in
+    // by /support/runtime. We need to run one more `go get`. We could have possibly frontloaded
+    // this work by adding a hand curated list of dependencies in project creation, but this
+    // would be more birttle.
+    const getDeps = spawn.sync("go", ["get"], {
+      cwd: this.sourceDir,
+      stdio: [/* stdin=*/ "ignore", /* stdout=*/ "pipe", /* stderr=*/ "pipe"],
+    });
+    logger.debug(getDeps.stdout.toString());
+    if (getDeps.status != 0) {
+      throw new FirebaseError("Failed to get dependencies", {
+        children: [new Error(getDeps.stderr.toString())],
+      });
     }
-
-    return Promise.resolve();
   }
 
+  // Watch isn't supported for Go
   watch(): Promise<() => Promise<void>> {
     return Promise.resolve(() => Promise.resolve());
   }
 
-  discoverSpec(
+  serve(
+    port: number,
+    adminPort: number,
+    envs: backend.EnvironmentVariables
+  ): Promise<() => Promise<void>> {
+    const childProcess = spawn("go", ["run", "./autogen"], {
+      env: {
+        ...envs,
+        ...process.env,
```

Comment:
Is inheriting `process.env` intended? I would've thought that we won't want to do this in anticipation of a future where parsing is done on the server side.

Reply:
It turns out the magic env is `HOME`. Without `HOME` set all `go` commands fail with the error `error missing $GOPATH`...